lib/archive/unpack_unix.go

```diff
@@ -172,8 +174,45 @@ func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root st
 	}
 	done := make(chan error)
 	go func() {
-		// copy the tarStream to the binary via stdin; the binary will stream it to InvokeUnpack unchanged
 		defer stdin.Close()
+		if tr, ok := tarStream.(*tar.Reader); ok {
```

Comment:
By what path do we get here? I suspect we're going `bytestream->tarstream->bytestream->tarstream` and we may have another option that is to suppress the first `bytestream->tarstream` translation.

Reply:
The following only applies to the online case.

The flow is tough to follow but the short answer is: we only actually copy the tarstream once, and it's tarstream->tarstream, over stdin.

The long answer is that the PortLayer sends the data to vSphere as SOAP data one way or the other (I didn't dig too deeply into how *that* works), and the govmomi code pulls it out of vSphere (again, didn't dig in because I didn't want to modify the toolbox) and expects a handling function that takes a `tar.Reader`, so by the time that my code gets the data in the toolbox, in `toolboxOverrideArchiveRead` down below, the data is already a tarstream. It's passed by reference to `Unpack`, so no copy there. Then it's streamed to the binary over stdin -- the binary calls `InvokeUnpack` with its `stdin` as the reference that's passed directly to `InvokeUnpack`, so you can think of it as streaming from `Unpack` to `InvokeUnpack` with a single copy and no data transformation to a generic bytestream for the Online case. `InvokeUnpack` is written to expect either a `tar.Reader` or a generic `io.Reader` because the Offline case produces the latter.